File: fli/search/_urls.py
"""URL-construction helpers for the FlightsFrontendService RPC endpoints.

Currently the only knob we expose at the URL layer is the locale tuple
(``curr=``, ``hl=``, ``gl=``). Google honours these on every endpoint we
talk to and they materially change the prices / language of the response,
so we surface them as explicit kwargs on the search methods rather than
hiding them as undocumented HTTP details.
"""

from __future__ import annotations

import urllib.parse


def with_locale_params(
    url: str,
    currency: str | None,
    language: str | None,
    country: str | None,
) -> str:
    """Append optional ``curr``/``hl``/``gl`` parameters to ``url``.

    - ``currency`` is uppercased ("usd" → "USD") because Google rejects
      lowercase codes silently (still 200, but ignores the override).
    - ``language`` is passed through verbatim (BCP-47, may contain a hyphen).
    - ``country`` is uppercased (ISO 3166-1 alpha-2).

    All values are percent-encoded so a caller typo (a stray ``&``, ``+``,
    or whitespace) can't break the URL — typical inputs like ``en-GB`` are
    already URL-safe so the encoding is a no-op for them.

    No-op when all three are None — returns the input URL unchanged so
    callers can pass through without checking.
    """
    params: list[str] = []
    if currency:
        params.append(f"curr={urllib.parse.quote(currency.upper(), safe='')}")
    if language:
        params.append(f"hl={urllib.parse.quote(language, safe='')}")
    if country:
        params.append(f"gl={urllib.parse.quote(country.upper(), safe='')}")
    if not params:
        return url
    sep = "&" if "?" in url else "?"
    return f"{url}{sep}{'&'.join(params)}"

File: fli-js/src/search/urls.ts
/**
 * URL helpers for the FlightsFrontendService RPCs.
 * 1:1 port of fli/search/_urls.py.
 */

/**
 * Append optional `curr`/`hl`/`gl` parameters to a URL.
 *
 * - `currency` is uppercased ("usd" → "USD") because Google rejects lowercase
 *   codes silently.
 * - `language` is passed through verbatim (BCP-47).
 * - `country` is uppercased (ISO 3166-1 alpha-2).
 *
 * All values are percent-encoded; returns `url` unchanged when all three are null.
 */
export function withLocaleParams(
  url: string,
  currency: string | null | undefined,
  language: string | null | undefined,
  country: string | null | undefined,
): string {
  const params: string[] = [];
  if (currency) params.push(`curr=${encodeURIComponent(currency.toUpperCase())}`);
  if (language) params.push(`hl=${encodeURIComponent(language)}`);
  if (country) params.push(`gl=${encodeURIComponent(country.toUpperCase())}`);
  if (params.length === 0) return url;
  const sep = url.includes("?") ? "&" : "?";
  return `${url}${sep}${params.join("&")}`;
}
